Python/_warnings.c

```diff
@@ -765,6 +899,151 @@ warn_explicit(PyThreadState *tstate, PyObject *category, PyObject *message,
+        if (module == NULL)
+            return NULL;
+    }
+    else
+        Py_INCREF(module);
+
+    /* Normalize message. */
+    Py_INCREF(message);  /* DECREF'ed in cleanup. */
+    rc = PyObject_IsInstance(message, PyExc_Warning);
+    if (rc == -1) {
+        goto cleanup;
+    }
+    if (rc == 1) {
+        text = PyObject_Str(message);
+        if (text == NULL)
+            goto cleanup;
+        category = (PyObject*)Py_TYPE(message);
+    }
+    else {
+        text = message;
+        message = PyObject_CallOneArg(category, message);
+        if (message == NULL)
+            goto cleanup;
+    }
+
+    lineno_obj = PyLong_FromLong(lineno);
+    if (lineno_obj == NULL)
+        goto cleanup;
+
+    if (source == Py_None) {
+        source = NULL;
+    }
+
+    /* Create key. */
+    key = PyTuple_Pack(4, text, fix_txt, category, lineno_obj);
+    if (key == NULL)
+        goto cleanup;
+
+    if ((registry != NULL) && (registry != Py_None)) {
+        rc = already_warned(interp, registry, key, 0);
+        if (rc == -1)
+            goto cleanup;
+        else if (rc == 1)
+            goto return_none;
+        /* Else this warning hasn't been generated before. */
+    }
+
+    action = get_filter(interp, category, text, lineno, module, &item);
+    if (action == NULL)
+        goto cleanup;
+
+    if (_PyUnicode_EqualToASCIIString(action, "error")) {
+        PyErr_SetObject(category, message);
+        goto cleanup;
+    }
+
+    if (_PyUnicode_EqualToASCIIString(action, "ignore")) {
+        goto return_none;
+    }
+
+    /* Store in the registry that we've been here, *except* when the action
+       is "always". */
+    rc = 0;
+    if (!_PyUnicode_EqualToASCIIString(action, "always")) {
+        if (registry != NULL && registry != Py_None &&
+            PyDict_SetItem(registry, key, Py_True) < 0)
+        {
+            goto cleanup;
+        }
+
+        if (_PyUnicode_EqualToASCIIString(action, "once")) {
+            if (registry == NULL || registry == Py_None) {
+                registry = get_once_registry(interp);
+                if (registry == NULL)
+                    goto cleanup;
+            }
+            /* WarningsState.once_registry[(text, category)] = 1 */
+            rc = update_registry(interp, registry, text, category, 0);
+        }
+        else if (_PyUnicode_EqualToASCIIString(action, "module")) {
+            /* registry[(text, category, 0)] = 1 */
+            if (registry != NULL && registry != Py_None)
+                rc = update_registry(interp, registry, text, category, 0);
+        }
+        else if (!_PyUnicode_EqualToASCIIString(action, "default")) {
+            PyErr_Format(PyExc_RuntimeError,
+                        "Unrecognized action (%R) in warnings.filters:\n %R",
+                        action, item);
+            goto cleanup;
+        }
+    }
+
+    if (rc == 1)  /* Already warned for this module. */
+        goto return_none;
+    if (rc == 0) {
```

Comment:
Would this be easier as `if (rc == -1 || call_show_warning_with_fix(...) < 0)`?

Reply:
Makes sense mostly, will combine, but am curious why `rc == -1` instead of `rc == 0`

Reply:
It depends what values `rc` can actually return (I think it's only -1, 0, +1). If you want to be explicit you can do `if (rc == -1 || (rc == 0 && call_show_warning_with_fix(...))`.

Reply:
Fixed in commit [fdc0086](https://github.com/softdevteam/pygrate3/pull/5/commits/fdc00867136da935fabd3325b48145fa9a1e757a)

Reply:
Referenced the wrong commit, my bad,

Reply:
I think https://github.com/softdevteam/pygrate3/pull/5/commits/baa4692a74f3212b268a5385dd2dcf87130f545c has changed the logic. Now if `rc==0` then `goto cleanup` will be executed unconditionally without calling `call_show_warning_with_fix`.

Reply:
Subinterpreter states were malfunctioning too. I have used your suggestions in commit [e0ee5f3](https://github.com/softdevteam/pygrate3/pull/5/commits/e0ee5f346fbb7ad715ea29cee7740b287255a402)